internal/rbd/rbd_util.go

```diff
@@ -417,25 +417,35 @@ func (rv *rbdVolume) getCloneDepth(ctx context.Context) (uint, error) {
 	}
 }
 
-func flattenClonedRbdImages(ctx context.Context, snaps []snapshotInfo, pool, monitors string, cr *util.Credentials) error {
-	rv := &rbdVolume{
-		Monitors: monitors,
-		Pool:     pool,
-	}
-	defer rv.Destroy()
-	err := rv.Connect(cr)
-	if err != nil {
-		util.ErrorLog(ctx, "failed to open connection %s; err %v", rv, err)
-		return err
-	}
-	for _, s := range snaps {
-		if s.Namespace.Type == "trash" {
-			rv.RbdImageName = s.Namespace.OriginalName
-			err = rv.flattenRbdImage(ctx, cr, true, rbdHardMaxCloneDepth, rbdSoftMaxCloneDepth)
-			if err != nil {
-				util.ErrorLog(ctx, "failed to flatten %s; err %v", rv, err)
-				continue
+type trashSnapInfo struct {
+	origSnapName string
+}
+
+func (rv *rbdVolume) flattenClonedRbdImages(ctx context.Context, snaps []librbd.SnapInfo, cr *util.Credentials) error {
+	var origNameList []trashSnapInfo
+	for _, snapInfo := range snaps {
+		// check if the snapshot belongs to trash namespace.
+		isTrash, retErr := rv.isTrashSnap(snapInfo.Id)
+		if retErr != nil {
+			return retErr
+		}
+
+		if isTrash {
+			// get original snap name for the snapshot in trash namespace
+			origSnapName, retErr := rv.getOrigSnapName(snapInfo.Id)
+			if retErr != nil {
+				return retErr
 			}
+			origNameList = append(origNameList, trashSnapInfo{origSnapName})
+		}
+	}
+
+	for _, snapName := range origNameList {
+		rv.RbdImageName = snapName.origSnapName
```

Comment:
This looks dangerous and might cause unexpected behaviour. Please do not change `RbdImageName` here.

Possible options:
- copy `rv` and use that
- pass the `snapName.origSnapName` to `rv.flattenRbdImage()`

Reply:
Sorry, it was a mistake.

What if I use the function like this (without pointer), as suggested by @Madhu-1 :

func **(rv rbdVolume)** flattenClonedRbdImage

Reply:
Please ignore my previous comment as that won't work because of following error:

internal/rbd/rbd_util.go:424:7: hugeParam: rv is heavy (408 bytes); consider passing it by pointer (gocritic)
func (rv rbdVolume) flattenClonedRbdImages(ctx context.Context, snaps []librbd.SnapInfo, cr *util.Credentials) error {

I guess I should restore to the previous state where we were creating a local rbdVolume and passing only few things.